code/datums/hud.dm

```diff
@@ -17,116 +21,356 @@ GLOBAL_LIST_INIT(huds, list(
+	var/list/hud_icons = list()
+
+	///mobs associated with the next time this hud can be added to them
+	var/list/next_time_allowed = list()
+	///mobs that have triggered the cooldown and are queued to see the hud, but do not yet
+	var/list/queued_to_see = list()
+	/// huduser = list(atoms with their hud hidden) - aka everyone hates targeted invisiblity
+	var/list/hud_exceptions = list()
+	///whether or not this atom_hud type updates the global huds_by_category list.
+	///some subtypes cant work like this since theyre supposed to "belong" to
+	///one target atom each. it will still go in the other global hud lists.
+	var/uses_global_hud_category = TRUE
 
 /datum/atom_hud/New()
 	GLOB.all_huds += src
+	for(var/z_level in 1 to world.maxz)
+		hud_atoms += list(list())
+		hud_users += list(list())
+
+	RegisterSignal(SSdcs, COMSIG_GLOB_NEW_Z, .proc/add_z_level_huds)
+
+	if(uses_global_hud_category)
+		for(var/hud_icon in hud_icons)
+			GLOB.huds_by_category[hud_icon] += list(src)
 
 /datum/atom_hud/Destroy()
-	for(var/v in hudusers)
-		remove_hud_from(v)
-	for(var/v in hudatoms)
-		remove_from_hud(v)
+	for(var/mob/mob as anything in hud_users_all_z_levels)
+		hide_from(mob)
+
+	for(var/atom/atom as anything in hud_atoms_all_z_levels)
+		remove_atom_from_hud(atom)
+
+	if(uses_global_hud_category)
+		for(var/hud_icon in hud_icons)
+			LAZYREMOVEASSOC(GLOB.huds_by_category, hud_icon, src)
+
 	GLOB.all_huds -= src
 	return ..()
 
-/datum/atom_hud/proc/remove_hud_from(mob/M, absolute = FALSE)
-	if(!M || !hudusers[M])
-		return
-	if (absolute || !--hudusers[M])
-		UnregisterSignal(M, COMSIG_PARENT_QDELETING)
-		hudusers -= M
-		if(next_time_allowed[M])
-			next_time_allowed -= M
-		if(queued_to_see[M])
-			queued_to_see -= M
+/datum/atom_hud/proc/add_z_level_huds()
+	SIGNAL_HANDLER
+	hud_atoms += list(list())
+	hud_users += list(list())
+
+///returns a list of all hud atoms in the given z level and linked lower z levels (because hud users in higher z levels can see below)
+/datum/atom_hud/proc/get_hud_atoms_for_z_level(z_level)
+	if(z_level <= 0)
+		return FALSE
+	if(z_level > length(hud_atoms))
+		stack_trace("get_hud_atoms_for_z_level() was given a z level index out of bounds of hud_atoms!")
+		return FALSE
+
+	. = list()
+	. += hud_atoms[z_level]
+
+	var/max_number_of_linked_z_levels_i_care_to_support_here = 10
+
+	while(max_number_of_linked_z_levels_i_care_to_support_here)
+		var/lower_z_level_exists = SSmapping.level_trait(z_level, ZTRAIT_DOWN)
+
+		if(lower_z_level_exists)
+			z_level--
+			. += hud_atoms[z_level]
+			max_number_of_linked_z_levels_i_care_to_support_here--
+			continue
+
 		else
-			for(var/atom/A in hudatoms)
-				remove_from_single_hud(M, A)
+			break
 
-/datum/atom_hud/proc/remove_from_hud(atom/A)
-	if(!A)
+///returns a list of all hud users in the given z level and linked upper z levels (because hud users in higher z levels can see below)
+/datum/atom_hud/proc/get_hud_users_for_z_level(z_level)
+	if(z_level <= 0)
```

Comment:
should combine this with the stack_trace() below

Reply:
yeet